README.md
# php-future

[![Build Status](https://travis-ci.org/resonantcore/php-future.svg?branch=master)](https://travis-ci.org/resonantcore/php-future)

This library backports some new features into old versions of PHP.

This **MIT Licensed** project is produced and maintained by [Resonant Core](https://resonantcore.net) for the good of the PHP developer community.

## WARNING ABOUT APPLICATION SECURITY

This should ONLY be used in the case that you cannot upgrade PHP for specific reasons within your business rules. For best application security you should always run the latest version of PHP where ever possible. This is intended as a stop gap for those that cannot upgrade and can be used while working toward upgrading to the latest version.

## ATTENTION PHP 5.2 / 5.3 USERS

As a general rule, we do not provide support for [any unsupported versions of PHP](http://php.net/eol.php). This means 5.4.x as of this writing.

If your operating system supports a 5.3.x version that is EOL'd (e.g. PHP 5.3.10 on Ubuntu 12.04 has security fixes backported), we will make exceptions. **The purpose of this library is not to encourage complacency and insecurity. Patch your systems.**

We do not support PHP 5.2.x anywhere. Please upgrade to a newer version of PHP.

## Features polyfilled by this library

* PHP 5.4
  * N/A
* PHP 5.5
  * `array_column()`
  * `boolval()`
  * `hash_pbkdf2()`
  * `openssl_pbkdf2()`
* PHP 5.6
  * `hash_equals()`
* PHP 7.0
  * N/A

See [wishlist.txt](wishlist.txt) for features we will be providing as time goes on. **To add to this list, either send a pull request, a [tweet](https://twitter.com/ResonantCore), or [an email](https://resonantcore.net/contact) and we'll consider it for inclusion.**

### Out of scope:

* Password API - Go directly to https://github.com/ircmaxell/password_compat for these implementations
  * `password_hash()`
  * `password_verify()`

## Why?

When our consultants submit security enhancements to open source projects, the
maintainers sometimes complain that some features are only in newer versions of 
PHP so they can't require them. We wrote this to provide compatibility across
all versions of PHP and prevent these issues from occurring.

In the future, we will attempt to backport *some* PHP 7 features for use in
versions 5.5 and 5.6, where possible.

## How it Works

Features that only exist in future versions are exposed in the global namespace.
These functions call the appropriate classes/methods from the `src` folder. This
ensures that PHP only load the codes when it actually needs it.

## How to use this library

### Composer

```sh
composer require resonantcore/php-future
```

### Any script

Just include our `autoload.php` script outside of our `src` directory and you're
good to go. For example:

```php
require_once "vendor/resonantcore/php-future/autoload.php";
```


CHANGELOG.md
# Changelog

## 0.3.0

Widened scope to include 5.3 because of LTS operating systems.
Also added a `BaseFuture` class to prevent code duplication.

### New functions polyfilled

* **PHP 5.4**
    * `getimagesizefromstring()`
    * `hex2bin()`

### New interfaces polyfilled

* **PHP 5.4**
    * `JsonSerializable`
    * `SessionHandlerInterface`

## 0.2.0

### New functions polyfilled

* **PHP 5.5**
    * `array_column()`
    * `boolval()`
    * `hash_pbkdf2()`
    * `openssl_pbkdf2()`

## 0.1.1

Added unit test coverage.

## 0.1.0

Initial release. Just polyfilled hash_equals()

tests/Utility.php
<?php
use \ResonantCore\PHPFuture as Future;

class TestSecurity extends PHPUnit_Framework_TestCase
{
    /**
     * @covers \ResonantCore\PHPFuture\Utilty::arrayColumn()
     */
    public function testArrayColumn()
    {
        $x = [
            ['a' => 1, 'b' => 1],
            ['a' => 2, 'b' => 3],
            ['a' => 3, 'b' => 2]
        ];

        $y = [
            1 => 1,
            2 => 3,
            3 => 2
        ];

        $this->assertEquals(
            $y,
            Future\Utility::arrayColumn($x, 'b', 'a')
        );
    }
}


tests/Security.php
<?php
use \ResonantCore\PHPFuture as Future;

class TestSecurity extends PHPUnit_Framework_TestCase
{
    /**
     * @covers \ResonantCore\PHPFuture\Security::hashEquals()
     */
    public function testHashEquals()
    {
        $a = 'd4ef98335dc5512500ea74a70a26dbd2759acc0ac116f0d747d05848a8365da9';
        $b = 'd4ef98335dc5512500ea74a70a26dbd2759acc0ac116f0d747d05848a8365da9';
        
        // Obvious test case:
        $this->assertTrue(\hash_equals($a, $b));
        
        $a = 'd4ef98335dc5512500ea74a70a26dbd2759acc0ac116f0d747d05848a8365da9';
        $b = 'd4ef98335dc5512500ea74a70a26dbd2759acc0ac116f0d747d05848a8365da1';
        
        $this->assertFalse(\hash_equals($a, $b));
    }

    /**
     * @covers \ResonantCore\PHPFuture\Security::pbkdf2()
     * ref https://www.ietf.org/rfc/rfc6070.txt
     */
    public function testPBKDF2()
    {
        $a = Future\Security::pbkdf2("sha1", "password", "salt", 2, 20, true);
        $this->assertEquals(
            $a,
            hex2bin('ea6c014dc72d6f8ccd1ed92ace1d41f0d8de8957')
        );

        $b = Future\Security::pbkdf2("sha1", "password", "salt", 2, 20, false);
        $this->assertEquals(
            $b,
            'ea6c014dc72d6f8ccd1ed92ace1d41f0d8de8957'
        );

        
        
    }
}


src/Image.php
<?php

namespace ResonantCore\PHPFuture;

/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 Resonant Core, LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
class Image extends BaseFuture
{
    /**
     * Get the size of an image from a string
     *
     * @param string $imagedata - Binary image data
     * @param &array $imageinfo - Reference to a metadata array
     *
     * @return array
     */
    public static function imageSizeFromString($imagedata, &$imageinfo = null)
    {
        $uri = 'data://application/octet-stream;base64,'.\base64_encode($imagedata);
        if ($imageinfo !== null) {
            return \getimagesize($uri, $imageinfo);
        }
        return \getimagesize($uri);
    }
}


src/Utility.php
<?php

namespace ResonantCore\PHPFuture;

/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 Resonant Core, LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
class Utility extends BaseFuture
{
    /**
     * Return the values from a single column in the input array
     *
     * @param array $array A multi-dimensional array (record set) from which to pull a column of values.
     * @param string|int $column_key The column of values to return.
     * @param string|int|null $index_key The column to use as the index/keys for the returned array.
     *
     * @return array
     */
    public static function arrayColumn(array $array, $column_key, $index_key = null)
    {
        $aReturn = array;
        if ($column_key === null) {
            // No column key? Grab the whole row...
            if ($index_key === null) {
                return $array;
            }
            foreach ($array as $sub) {
                $aReturn[$sub[$index_key]] = $sub;
            }
        } elseif (empty($index_key)) {
            foreach ($array as $sub) {
                $aReturn[] = $sub[$column_key];
            }
        } else {
            foreach ($array as $sub) {
                $aReturn[$sub[$index_key]] = $sub[$column_key];
            }
        }
        return $aReturn;
    }

    /**
     * Convert a hexadecimal string into raw binary
     *
     * @param string $data Hexadecimal data
     *
     * @return string
     */
    public static function hexToBin($data)
    {
        if (self::ourStrlen($data) % 2 !== 0) {
            \trigger_error("hex2bin(): Hexadecimal input string must have an even length", E_USER_WARNING);
            return false;
        }
        return \pack('H*', $data);
    }
}


src/Security.php
<?php
namespace ResonantCore\PHPFuture;

/**
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015 Resonant Core, LLC
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
class Security extends BaseFuture
{
    /**
     * Equivalent to hash_equals() in PHP 5.6 
     * 
     * @param  string $knownString
     * @param  string $userString
     * @return bool
     */
    public static function hashEquals($knownString, $userString)
    {
        // We have to roll our own
        $kLen = self::ourStrlen($knownString);
        $uLen = self::ourStrlen($userString);
        if ($kLen !== $uLen) {
            return false;
        }
        $result = 0;
        for ($i = 0; $i < $kLen; $i++) {
            $result |= (ord($knownString[$i]) ^ ord($userString[$i]));
        }
        // They are only identical strings if $result is exactly 0...
        return 0 === $result;
    }
    
    /**
     * Password Based Key Derivation Function #2
     *
     * Underlying feature for hash_pbkdf2() and openssl_pbkdf2() for PHP < 5.5
     *
     * @param string $algo
     * @param string $password
     * @param string $salt
     * @param int $iterations
     * @param int $length
     */
    public static function pbkdf2($algo, $password, $salt, $iterations, $length = 0)
    {
        if (!\in_array($algo, \hash_algos(), true)) {
            throw new \Exception('PBKDF2 ERROR: Invalid hash algorithm.');
        }

        $hashlen = self::ourStrlen(\hash($algo, "", true));
        if ($hashlen === 0) {
            return '';
        }

        $block_count = \ceil($length / $hashlen);

        $output = '';
        for ($i = 1; $i <= $block_count; ++$i) {
            $last = $salt.\pack('N', $i);
            $xorsum = \hash_hmac($algo, $last, $password, true);
            $last = $xorsum;
            for ($j = 1; $j < $iterations; ++$j) {
                $xorsum ^= ($last = \hash_hmac($algorithm, $last, $password, true));
            }
            $output .= $xorsum;
        }
        return self::ourSubstr($output, 0, $length);
    }
}


src/BaseFuture.php
<?php
namespace Resonantcore\PHPFuture;

/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 Resonant Core, LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
class BaseFuture
{
    /**
     * Multi-byte-safe string length calculation
     *
     * @param string $str
     * @return int
     */
    protected static function ourStrlen($str)
    {
        // Premature optimization: cache the function_exists() result
        static $exists = null;
        if ($exists === null) {
            $exists = \function_exists('\\mb_strlen');
        }

        // If it exists, we need to make sure we're using 8bit mode
        if ($exists) {
            return \mb_strlen($str, '8bit');
        }
        return \strlen($str);
    }

    /**
     * Multi-byte-safe substring calculation
     *
     * @param string $str
     * @param int $start
     * @param int $length (optional)
     * @return string
     */
    protected static function ourSubstr($str, $start = 0, $length = null)
    {
        // Premature optimization: cache the function_exists() result
        static $exists = null;
        if ($exists === null) {
            $exists = \function_exists('\\mb_substr');
        }

        // If it exists, we need to make sure we're using 8bit mode
        if ($exists) {
            return \mb_substr($str, $start, $length, '8bit');
        } elseif ($length !== null) {
            return \substr($str, $start, $length);
        }
        return \substr($str, $start);
    }
}


autoload.php
<?php
use \Resonantcore\PHPFuture as Future;
spl_autoload_register(function ($class) {
    
    // project-specific namespace prefix
    $prefix = 'Resonantcore\\PHPFuture\\';

    // base directory for the namespace prefix
    $base_dir = __DIR__ . '/src/';
    
    // does the class use the namespace prefix?
    $len = strlen($prefix);
    if (strncmp($prefix, $class, $len) !== 0) {
        // no, move to the next registered autoloader
        return;
    }
    
    // get the relative class name
    $relative_class = substr($class, $len);
    
    // replace the namespace prefix with the base directory, replace namespace
    // separators with directory separators in the relative class name, append
    // with .php
    $file = $base_dir . str_replace('\\', '/', $relative_class) . '.php';
    
    // if the file exists, require it
    if (file_exists($file)) {
        require $file;
    }
});


if (!interface_exists('JsonSerializable')) {
    /**
     * From PHP 5.4
     *
     * @ref http://us3.php.net/manual/en/class.jsonserializable.php
     */
    interface JsonSerializable
    {
        public function jsonSerialize();
    }
}

if (!interface_exists('SessionHandlerInterface')) {

    /**
     * From PHP 5.4
     *
     * @ref http://php.net/manual/en/class.sessionhandlerinterface.php
     */
    interface SessionHandlerInterface
    {
        public function close();
        public function destroy($session_id);
        public function gc($maxlifetime);
        public function open($save_path, $name);
        public function read($session_id);
        public function write($session_id, $session_data);
    }
}


if (!function_exists('array_column')) {
    /**
     * Get the boolean value of a variable
     */
    function array_column(array $array, $column_key, $index_key = null)
    {
        return Future\Utility::arrayColumn($array, $column_key, $index_key);
    }
}

if (!function_exists('boolval')) {
    /**
     * Get the boolean value of a variable
     */
    function boolval($mixed_var)
    {
        return !!$mixed_var;
    }
}

if (!function_exists('hex2bin')) {
    /**
     * From PHP 5.4
     *
     * @ref https://php.net/hex2bin
     *
     * @param string $data
     *
     * @return string
     */
    function hex2bin($data)
    {
        return Future\Utility::hexToBin($data);
    }
}

if (!function_exists('hash_equals')) {
    /**
     * From PHP 5.6
     * 
     * @ref https://php.net/hash_equals
     *
     * @param string $a
     * @param string $b
     *
     * @return boolean
     */
    function hash_equals($known_string, $user_string)
    {
        return Future\Security::hashEquals(
            $known_string,
            $user_string
        );
    }
}

if (!function_exists('hash_pbkdf2')) {
    /**
     * From PHP 5.5
     *
     * @ref https://php.net/hash_pbkdf2
     *
     * @param string $algo
     * @param string $password
     * @param string $salt
     * @param int $iterations
     * @param int $length
     * @param boolean $raw_output
     *
     * @return string
     */
    function hash_pbkdf2($algo, $password, $salt, $iterations, $length = 0, $raw_output = false)
    {
        $key = Future\Security::pbkdf2(
            $algo,
            $password,
            $salt,
            $iterations,
            $length
        );
        if ($raw_output) {
            return bin2hex($key);
        }
        return $key;
    }
}

if (!function_exists('getimagesizefromstring')) {
    /**
     * From PHP 5.4
     *
     * @ref https://php.net/getimagesizefromstring
     *
     * @param string $imagedata
     * @param &array $imageinfo
     *
     * @return array
     */
    function getimagesizefromstring($imagedata, &$imageinfo = null)
    {
        if ($imageinfo !== null) {
            return Future\Image::imageSizeFromString($imagedata, $imageinfo);
        }
        return Future\Image::imageSizeFromString($imagedata);
    }
}

if (!function_exists('openssl_pbkdf2')) {
    /**
     * From PHP 5.5
     *
     * @ref https://php.net/openssl_pbkdf2
     *
     * @param string $password
     * @param string $salt
     * @param int $iterations
     * @param int $length
     * @param string $algo
     *
     * @return string
     */
    function openssl_pbkdf2($password, $salt, $length, $iterations, $algo = 'sha1')
    {
        $key = Future\Security::pbkdf2(
            $algo,
            $password,
            $salt,
            $iterations,
            $length
        );
        return $key;
    }
}
